Print Contract Allocations

bh2smith opened the pull request:
Tiny script to print airdrop allocations given to smart contracts that are **not** wallets.



Depends on #14 

Test Plan:

```sh
 python -m src.scripts.contract_allocations
```

Will print the following:

```
loading contracts from mainnet-contracts.txt
making batch call for 357 addresses on mainnet (this will take a while)...
found 278 wallet contracts on mainnet
found 49 contract allocations to non-wallets on mainnet.
printing Top 3
1. 0x7f30a53f290fc551a089bd83107906be649a92bd - 1.104M
2. 0xbf74c12d6e6a8c2bd7f36a917250e2bf8bffefc1 - 0.348M
3. 0x3e8468f66d30fc99f745481d4b383f89861702c6 - 0.010M
total mainnet (in WEI) 1494936942894087768984187
loading contracts from gchain-contracts.txt
making batch call for 64 addresses on gchain (this will take a while)...
found 47 wallet contracts on gchain
found 16 contract allocations to non-wallets on gchain.
printing Top 3
1. 0x918a1f8ee8f8f9e0802071929ce5fe9eff60ac8c - 1.622M
2. 0x647507a70ff598f386cb96ae5046486389368c66 - 1.330M
3. 0xf6a78083ca3e2a662d6dd1703c939c8ace2e268d - 0.306M
total gchain (in WEI) 3276998369577464636110864

Grand Total Non Wallet Airdrop Allocation (in WEI) 4771935312471552405095051
```

nlordell commented:
Ran it and it looks good to me.

Note I had to generate the main allocations before running this `python -m src.main`.

I did not look over all of the contracts but I checked a few of them and they looked correct. Also, some of the unverified contracts are hard to determine whether or not they are wallets. With that in mind - I think we have to consider all unverified contracts as wallets. Otherwise, they could claim and swap for COW and then we won't have enough COW balance in the contract to cover all the swaps.

We should only discount contracts that we know for absolute sure cannot swap COW if claimed (which required verified code).

PS: **reminder** - remove the print statements.

bh2smith commented:
Thanks @nlordell - I have cleaned up a bit, applied your suggestion regarding unverified contracts and made everything extra nice. 

Updated the PR description with deterministic logs for this script.

bh2smith committed:
script to determine non wallet allocations

bh2smith committed:
rework contract class for batch calls

bh2smith committed:
unused import

bh2smith committed:
Merge branch 'pre-contract-allocations' into contract-allocations

bh2smith committed:
missing import

bh2smith committed:
only report Airdrop category

MartinquaXD commented on src/scripts/contract_allocations.py:
```diff
@@ -24,21 +24,21 @@ def non_wallet_allocation(network: str, allocation_file: File) -> int:
-    for allocation in sorted_contract_allocations:
-        if allocation.Account in not_wallets:
-            gno_allocation = allocation.Airdrop + allocation.GnoOption
+    airdrop_total, found = 0, 0
```
Not sure if renaming `allocation` to `airdrop` in this and the following cases since it's about the sum of airdrop plus gno allocation.
I don't have strong feelings against the new name, it's just that it makes it harder to see what actually changed in the diff. (Which seem to be only print statements)

MartinquaXD reviewed (approved):
~~Changes seem fine.~~
Please ignore. I looked at an old version. Will leave another review.

MartinquaXD commented on src/fetch/contracts.py:
```diff
@@ -107,6 +110,40 @@ def contracts(self, load_from: NetworkFile) -> dict[str, bool]:
+                wallets |= set(accounts)
+                total_wallets += len(accounts)
+            elif byte_code in NOT_WALLET_BYTECODE:
+                not_wallets[NOT_WALLET_BYTECODE[byte_code]] = accounts
```
Seems like this case can be dropped entirely. If you want to keep it to not mess with the assertion in the `else` case a `pass` would be fine here, right?
In that case `not_wallets` can also be dropped.

bh2smith replied:
I think the idea was that if someone wanted more verbose logging about what's what, they could use these and add print statements. However, in its current state I completely agree it can be removed. I'm a bit hesitant though because I like the idea of seeing more details here. What do you think?

MartinquaXD replied:
I think optimizing the code for readability makes more sense than optimizing for people **maybe** needing more info. It's not difficult to add log info when you are debugging something.

bh2smith commented:
> ~~Changes seem fine.~~
> 
> Please ignore. I looked at an old version. Will leave another review.

Also note that I have split this into two PRs to make this easier to review.

MartinquaXD commented on src/fetch/contracts.py:
```diff
@@ -107,6 +110,40 @@ def contracts(self, load_from: NetworkFile) -> dict[str, bool]:
+        total_wallets = 0
+        wallets = set()
+        not_wallets = {}
+        unverified = {}
```
It doesn't look like the mapping of code to accounts in `unverified` gets used (i.e. look up) at any point.
In that case `unverified` could probably be a set which would simplify the important parts of the loop to this:
```python
            if byte_code in WALLET_BYTECODE:
                wallets |= set(accounts)
            elif byte_code in UNVERIFIED_BYTECODE:
                unverified |= set(accounts)
            # ...
                
            total_wallets = len(wallets | unverified)
```

bh2smith replied:
Right. It used to be used but this has been removed. Thanks.

MartinquaXD commented on src/scripts/contract_allocations.py:
```diff
@@ -0,0 +1,55 @@
+    )
+    contracts = account_info.contracts(load_from=NetworkFile("contracts.txt"))
+    not_wallets = account_info.get_non_wallets(
+        [c for c, is_contract in contracts.items() if is_contract]
```
Depending on how many contracts there are it might be worth it to put `contracts.items()` in a set to speed up the lookup here. But in the grand scheme of things it probably doesn't matter.

bh2smith replied:
I actually want to refactor how contracts are stored. Nowhere yet in this project have we ever used the false case so yes, a set would be a more appropriate return type with equivalent information. However I would prefer to do this in a separate PR.

bh2smith replied:
I'll create an issue for this and start on it ASAP.

bh2smith replied:
https://github.com/gnosis/cow-token-allocation/issues/15

MartinquaXD commented on src/scripts/contract_allocations.py:
```diff
@@ -0,0 +1,55 @@
+            gno_allocation = alloc.Airdrop + alloc.GnoOption
+            assert alloc.total() - gno_allocation == 0
+            airdrop_total += alloc.Airdrop
+            if found < 3:
+                found += 1
+                print(
+                    f"{found}. {alloc.Account} - {(alloc.Airdrop / 1e24):.3f}M"
+                )
```
This seems like a magic number to me. I'm guessing there are 3 "special" addresses which we are interested here.
To make it to read for people without background knowledge something like this could make sense:
```python
special_addresses = set(
    "0x...1", # reason why this is special
    # ...
)
for alloc in sorted_contract_allocations:
    if alloc.Account in not_wallets:
        if alloc.Account in special_addresses:
            print(
                f"{found}. {alloc.Account} - {(alloc.Airdrop / 1e24):.3f}M"
            )
        # ... rest of the loop

```

bh2smith replied:
They aren't really special, they are just the top 3. The print statement above says printing top three.

MartinquaXD replied:
Oh, I missed that. In that case ignore what I said. 🙈

josojo commented:
For the testplan, which allocation file should we use? The one within the repo?

Then I get:
```
python -m src.scripts.contract_allocations

loading contracts from mainnet-contracts.txt
making batch call for 391 addresses on mainnet (this will take a while)...
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.10/3.10.0_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/Cellar/python@3.10/3.10.0_2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/Users/alexherrmann/gnosis/cow-token-allocation/src/scripts/contract_allocations.py", line 45, in <module>
    mainnet_total = non_wallet_allocation(
  File "/Users/alexherrmann/gnosis/cow-token-allocation/src/scripts/contract_allocations.py", line 15, in non_wallet_allocation
    not_wallets = account_info.get_non_wallets(
  File "/Users/alexherrmann/gnosis/cow-token-allocation/src/fetch/contracts.py", line 138, in get_non_wallets
    raise RuntimeError(
RuntimeError: Unclassified Contract Bytecode:
0x608060405273ffffffffffffffffffffffffffffffffffffffff600054167fa619486e0000000000000000000000000000000000000000000000000000000060003514156050578060005260206000f35b3660008037600080366000845af43d6000803e60008114156070573d6000fd5b3d6000f3fea265627a7a7231582099bf9ecf033b5a5cd7073e2e717749e94f49c79dbfc065d6204cb99db2ba11ab64736f6c634300050c0032
example address: 0x506109c8c43b1838a466f37dc3f57019fa0537ca
```
`0x506109c8c43b1838a466f37dc3f57019fa0537ca` seems to be a safe.

bh2smith commented:
> 0x506109c8c43b1838a466f37dc3f57019fa0537ca seems to be a safe.

~~Yes, for some reason I have been getting this error a couple of times as well. Its very strange because if you search the project for this bytecode, you will find that it is indeed already included. I will have to figure out what is happening here...~~

Yes, it appears I was working with an older version of the contracts file! Will run contract file generation from scratch and update this contract bytecode.

bh2smith committed:
remove superfluous variables in favour of simplicity

bh2smith committed:
remove "total wallets

bh2smith committed:
include one stray safe that got away

bh2smith committed:
use set instead of dict

bh2smith committed:
Merge branch 'main' into contract-allocations

bh2smith committed:
import grouping
